### PR TITLE
TraumaDamageRework

### PR DESCRIPTION
**Pull request by MaxIsJoe**
### Purpose
Part of #6272
Some changes to simplify trauma damage, some balancing, Removes random bleeding values, Reimplements proper gibbing for mobs, placeholder trauma healing logic on items and some fixes.

## Balance changes
To keep damage more consistent for players random bleeding values are no longer a thing and damage values have been lowered for internal organ damage to give players more time to react to get themselves fixed when surgery gets fixed.

Slash, Pierce and Burn damage on weapons also now share the base weapon damage instead of having their own separate ones. Damage gets slightly multiplied for traumatic damages when a player is in critical condition.

## Placeholder logic on existing healing items.
You can now heal slash and pierece damage using Bruise packs and Omnipents will heal burn damage. This does not stop internal bleeding.

## Reimplemented gibbing

Gibbing players was poorly implemented and caused silent issues. I've reimplemented and fixed gibbing so it can work on all mobs and not just the player.
All limbs fall on the ground when a player gets gibbed, and yes, no longer walking torso-less players.

### Changelog:
CL: [New] - Added in the ability for all mobs that use HealthV2 to be properly gibbed.
CL: [Improvement] - Removed random bleeding values and lowered internal bleeding values.
CL: [Improvement] - Trauma damage is now based on the weapon's damage.
CL: [Improvement] - Balance changes across all null rod weapons.
CL: [Fix] - Players can no longer walk around with a torsoless body.
CL: [Fix] - Fixed some rare cases where the game would silently error out body part dismemberment.
CL: [Fix] - Bleeding effects did not account for when the player would bleed more than 8 units of blood.

### CODE REVIEW

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs [hunk 1]
```diff
@@ -298,11 +304,16 @@ public enum BurnDamageLevels
...
 
-		public enum TramuticDamageTypes
+		[Flags]
+		public enum TramuticDamageTypes 
 		{
-			SLASH,
-			PIERCE,
-			BURN
+			SLASH  = 0,
+			PIERCE = 1,
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: I think you Set up the flags wrong?
Apparently you can't use zero, 

wasn't this supposed to be integrated with the damage list on body parts?

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs [hunk 2]
```diff
@@ -721,18 +750,17 @@ public void StopExternalBleeding()
 		/// </summary>
 		public void InternalBleedingLogic(AttackType attackType = AttackType.Internal, DamageType damageType = DamageType.Brute)
 		{
-			float damageToTake = UnityEngine.Random.Range(MinMaxInternalBleedingValues.x, MinMaxInternalBleedingValues.y);
 			if(currentInternalBleedingDamage >= maximumInternalBleedDamage)
 			{
-				BodyPart currentParent = ContainedIn;
+				BodyPart currentParent = GetParent();
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: this is here because you'll be never accessing the root body part controller since its internal bleeding, you would never be internally on the outermost body part

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs [hunk 3]
```diff
@@ -800,51 +841,47 @@ private void CheckBurnDamageLevels()
 		/// </summary>
 		private void AshBodyPart()
 		{
+			Logger.Log($"Ashing bodyPart -> {this.name}");
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: a LOg

File: UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs [hunk 4]
```diff
@@ -952,12 +973,15 @@ public void Harvest()
...
-			//TODO: Reimplement
-
-			//never destroy players!
-			_ = Despawn.ServerSingle(gameObject);
+			Death();
+			_ = SoundManager.PlayAtPosition(SingletonSOSounds.Instance.Slip, gameObject.transform.position, gameObject); //TODO: replace with gibbing noise
+			CirculatorySystem.Bleed(GetTotalBlood());
+			foreach(RootBodyPartContainer container in RootBodyPartContainers.ToArray())
+			{
+				container.RemoveLimbs();
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:967]: RemoveLimbs() 
destroys the root body part game object, it shouldn't do that since you need that to implant body parts into people

File: UnityProject/Assets/Scripts/Items/Implants/Organs/Lungs.cs [hunk 5]
```diff
@@ -181,57 +160,34 @@ private bool BreatheIn(GasMix breathGasMix, ReagentMix blood, float efficiency)
...
 		// Counterintuitively, in humans respiration is stimulated by pressence of CO2 in the blood, not lack of oxygen
 		// May want to change this code to reflect that in the future so people don't hyperventilate when they are on nitrous oxide
 		var inGas = GAS2ReagentSingleton.Instance.GetGasToReagent(requiredGas);
-		float bloodCap =  RelatedPart.bloodType.GetGasCapacity(RelatedPart.BloodContainer.CurrentReagentMix);
-		float foreignCap =  RelatedPart.bloodType.GetGasCapacityForeign( RelatedPart.BloodContainer.CurrentReagentMix);
-		var ratioNativeBlood =  bloodCap / ( bloodCap + foreignCap);
-		float bloodSaturation =  RelatedPart.BloodContainer[RelatedPart.requiredReagent] * ratioNativeBlood / bloodCap;
-
-		if (bloodSaturation >=RelatedPart.HealthMaster.CirculatorySystem.BloodInfo.BLOOD_REAGENT_SATURATION_OKAY)
+		var saturation = (toInhale[inGas] + blood[inGas]) / RelatedPart.bloodType.GetGasCapacity(blood);
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Items/Implants/Organs/Lungs.cs]: my changes got reverted!

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs (hunk 1 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: yeah, use the shifting technique to assign values, something like this:

````cs
[Flags]
public enum TramuticDamageTypes 
{
    None = 0,
    Slash = 1 << 0,
    Pierce = 2 << 1,
    Burn = 3 << 2
}
````

Also, why did you  create combined versions after Burn?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: > Also, why did you create combined versions after Burn?

Because for some reason unity doesn't allow you to combine two flags. not sure if it's a bug or what exactly but that's a workaround for it for now.

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs (hunk 3 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: lÖg

File: UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs (hunk 4 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:967]: Except you don't you completely lost your body and you're now nothing but gibs and blood painting the floor. You lost your entire body.

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs (hunk 1 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: combine where? In the dropdown in editor?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: Yep

Comment (Bod9001) [reply UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: I think you have to use the naughty attributes flag

File: UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs (hunk 4 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:967]: well, it's not Needed then if the body is going to be destroyed

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:967]: Actually no, destroying the body object itself is just a whole can of issues + this solution is the only way we can do proper gibbing without the game throwing errors or doing anything strange. I've tried going through each individual limb and remove them and tried many other solutions and they all were a pain to fix so this is the only way to do proper gibbing right now and I don't think we need to approach this differently since it works as intended and isn't 50+ lines long.

File: UnityProject/Assets/Scripts/Items/Implants/Organs/Lungs.cs (hunk 5 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Items/Implants/Organs/Lungs.cs]: Reverted them back

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs (hunk 1 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: Yep, Bod was right. It now works correctly and adjusted it to that.

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs (hunk 2 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: I have read this multiple times and I'm not following, what exactly is the problem here?

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs [hunk 6]
```diff
@@ -181,9 +187,9 @@ public partial class BodyPart
 		public float RadiationStacks => Damages[(int) DamageType.Radiation];
 
 
-		[HideInInspector] public float CurrentInternalBleedingDamage
-		{
-			get
+		[HideInInspector] public float CurrentInternalBleedingDamage
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: properties don't show in inspector, you don't need to hide them

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs [hunk 7]
```diff
@@ -868,3 +902,13 @@ private void UpdateIcons()
...
 }
+
+
+[Flags]
+public enum TramuticDamageTypes
+{
+	NONE = 0,
+	SLASH = 1 << 0,
+	PIERCE = 2 << 1,
+	BURN = 4 << 2
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: 4 ? e.e

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: idk it only decided to work after doing 4

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs (hunk 6 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: fixed

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs (hunk 2 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs]: well, you would never have internal bleeding on a limb? 
Outer most Body part, so therefore you would never get root body part as a parent, 
also this code here causes errors when it does

File: UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs (hunk 4 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:967]: just don't destroy the body root part controller , the rest is fine

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:967]: done. I think we're ready to merge now.

**Comment by MaxIsJoe** [comment]
This has been ready for quite some time, need any testing or something?

**Comment by corp-0** [comment]
> Move 'TramuticDamageTypes' into a named namespace.

File: UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/BodyPart.cs [hunk 8]
```diff
@@ -305,7 +305,6 @@ public virtual void AddBodyPartSlot(ItemSlot ItemSlot)
 		/// Removes the Body Part Item from the storage of its parent (a body part container or another body part)
 		/// Will check if the this body part causes death upon removal and will tint it's Item Sprite to the character's skinTone if allowed.
 		/// </summary>
-		[ContextMenu("Debug - Drop this body part.")]
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/BodyPart.cs:302]: why was this removed?

File: UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/RootBodyPartContainer.cs [hunk 9]
```diff
@@ -389,33 +388,39 @@ public virtual void SubBodyPartAdded(BodyPart implant)
...
+		public void TakeTraumaDamage(float damage, TramuticDamageTypes damageType)
 		{
-			foreach (var ContainsLimb in ContainsLimbs)
+			//Check if we do more than one type of trauma damage at once.
+			//If true, don't run the indivual case check.
+			//We also do a check for a body part's existence in case it got removed by the previous damage.
+			foreach(BodyPart limb in ContainsLimbs)
 			{
-				ContainsLimb.ApplyTraumaDamage(damage, BodyPart.TramuticDamageTypes.BURN);
+				if(damageType.HasFlag(TramuticDamageTypes.BURN) && damageType.HasFlag(TramuticDamageTypes.SLASH)
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/RootBodyPartContainer.cs]: I don't understand this check

Comment (corp-0) [reply UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/RootBodyPartContainer.cs]: take a look at this snippet I did, notice how the blocks are entered without any fancy stuff, only 3 ifs. Also notice how the flag enum shifting was done wrongly in your pr, my bad.

https://onecompiler.com/csharp/3x4233yje

**Review by corp-0 — CHANGES_REQUESTED** [review]

File: UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/BodyPart.cs (hunk 8 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/BodyPart.cs:302]: I was going to say "Because `DismemberBodyPartLogic()` seems to be the better function to hold this so we don't have issues where people remove torsos then complain about walking around torsoless even though they caused this issue themselves." but then I took a good look at it and I think we can keep that gib check inside `RemoveBodyPartThis()` and the other function that replaced it is no longer needed.

**Comment by MaxIsJoe** [comment]
Fixed merge conflict. Ready for one more round or reviews.

**Comment by MaxIsJoe** [comment]
Alright I think I've fixed everything that was missing due to the merge thing. idk if I missed anything but hopefully I didn't

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs [hunk 10]
```diff
@@ -544,26 +568,41 @@ public void ApplyTraumaDamage(float tramuaDamage, TramuticDamageTypes damageType
 			//We use dismember protection chance because it's the most logical value.
 			if(DMMath.Prob(SelfArmor.DismembermentProtectionChance * 100) == false)
 			{
-				if(damageType == TramuticDamageTypes.SLASH) { currentSlashCutDamage += tramuaDamage; }
-				if(damageType == TramuticDamageTypes.PIERCE) { currentPierceDamage += tramuaDamage; }
+				if (damageType == TramuticDamageTypes.SLASH) { currentSlashCutDamage += MultiplyTraumaDamage(tramuaDamage); }
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs:571]: just making sure, this check will only work if ``damageType`` has SLASH flag and no other. Is this what you want? BTW make the body of the ``if`` have its own line.

File: UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/BodyPart.cs (hunk 8 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/BodyPart.cs:302]: This decorator is to make the function available in unity editor, to debug, nothing more. No one should complain about losing limbs when opening editor debug tab and clicking the function.

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs [hunk 11]
```diff
@@ -818,27 +857,14 @@ private void AshBodyPart()
 					var organ = item.ItemObject?.GetComponent<BodyPart>();
 					if (organ != null)
 					{
-						if (organ.DeathOnRemoval)
+						if (organ.gibsEntireBodyOnRemoval)
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs:859]: is this a public variable/property in camelCase?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs:859]: no it should be private, fixing this.

File: UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs (hunk 10 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/Damage/BodyPartDamage.cs:571]: This only gets called with one damage type so there is no need to check for multiple flags. Slash is the default in case no argument was given.

File: UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/BodyPart.cs (hunk 8 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/BodyPart.cs:302]: The key word here is _no one should_ but people who are debugging this might accidently think this is intended behavior and complain about it. So I'm making the logic consistent even for the people who are debugging.

**Review by corp-0 — APPROVED** [review]

File: UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/BodyPart.cs [hunk 12]
```diff
@@ -320,9 +319,12 @@ public virtual void RemoveFromBodyThis()
...
 
 		/// <summary>
-		/// Checks if a body part has a condition or state that applies logic upon removal then executes it.
+		/// Checks if it's possible to remove this body part and runs any logic
+		/// required upon it's removal.
 		/// </summary>
-		private void BodyPartRemovalChecks()
+		/// <returns>True if allowed to remove. Flase if gibbing.</returns>
+		[ContextMenu("Debug - Drop this Body Part")]
+		private bool BodyPartRemovalChecks()
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/BodyPart.cs:319]: I think this has to be public to work with context menu. Did you test it works being private?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/BodyPart.cs:319]: misplaced context menu, whoops.

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/BodyPart.cs:319]: wait a minute, this thing isn't misplaced on my end. what le fok github?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/BodyParts/BodyPart.cs:319]: fixed it by editing the file directly from github itself.

**Comment by MaxIsJoe** [comment]
Fixed another merge conflict

**Comment by MaxIsJoe** [comment]
Alright fixed a compile error involving someone changing a public variable to camelCase,

File: UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs [hunk 13]
```diff
@@ -226,7 +226,7 @@ public HungerState CalculateHungerState()
 		/// </summary>
 		private List<Sickness> immunedSickness;
 
-		public PlayerScript playerScript;
+		public PlayerScript PlayerScriptOwner;
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:229]: Why did you rename this back to PlayerScriptOwner, PlayerScript is fine? We usually just keep common script variable names the same.

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:229]: Like I said on Discord. camelCase on a public variable and in the context of that script it makes sense; because we're getting the owner of that healthMaster script
